File: src/main/java/me/santio/utils/listeners/InventoryListener.java
package me.santio.utils.listeners;

import me.santio.utils.SantioUtils;
import me.santio.utils.inventories.CustomInventory;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.util.Consumer;

import java.util.Optional;

public class InventoryListener implements Listener {
    
    private final SantioUtils utils;
    public InventoryListener(SantioUtils utils) {
        this.utils = utils;
    }
    
    @EventHandler
    public void onInventoryClose(InventoryCloseEvent event) {
        Optional<CustomInventory> inventory = utils.getInventories().values()
                .stream()
                .filter((CustomInventory i) -> i.isOpen((Player) event.getPlayer()))
                .findFirst();
        
        event.getPlayer().removeMetadata("sutils_inventory_open", utils.getPlugin());
        inventory.ifPresent((i) -> {
            if (i.isDeleteOnClose()) i.delete();
        });
    }
    
    @EventHandler
    public void onInventoryClick(InventoryClickEvent event) {
        if (event.getClickedInventory() == null || !(event.getWhoClicked() instanceof Player)) return;
        Player player = (Player) event.getWhoClicked();
        
        // Find which inventory the player has open
        Optional<CustomInventory> inventory = utils.getInventories().values()
                .stream()
                .filter((CustomInventory i) -> i.isOpen(player))
                .findFirst();
    
        if (!inventory.isPresent()) return;
        
        // Get item
        ItemStack item = event.getClickedInventory().getItem(event.getSlot());
        if (item == null) return;
        ItemMeta meta = item.getItemMeta();
        if (meta == null) return;
        
        // Fetch event ID
        Consumer<InventoryClickEvent> itemEvent = inventory.get().getEvents().get(utils.NBTUtils.get(meta, "sutils_id"));
        if (itemEvent == null) return;
    
        // Run event
        itemEvent.accept(event);
    }
    
}

File: src/main/java/me/santio/utils/inventory/InventoryListener.kt
package me.santio.utils.inventory

import me.santio.utils.SantioUtils
import me.santio.utils.bukkit.AsyncUtils
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.inventory.InventoryOpenEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.metadata.FixedMetadataValue

object InventoryListener: Listener {

    @EventHandler
    private fun onInventoryClose(event: InventoryCloseEvent) {
        if (event.player !is Player) return
        val inventory = CustomInventory.getOpenInventory(event.player as Player) ?: return
        event.player.removeMetadata("inventory", SantioUtils.plugin!!)

        if (inventory.deleteOnClose() && inventory.isEmpty()) inventory.delete()
    }

    @EventHandler
    private fun onInventoryOpen(event: InventoryOpenEvent) {
        if (event.player !is Player) return
        val inventory = CustomInventory.get(event.inventory) ?: return

        event.player.setMetadata("inventory", FixedMetadataValue(SantioUtils.plugin!!, inventory.id))
    }

    @EventHandler
    private fun onInventoryClick(event: InventoryClickEvent) {
        if (event.whoClicked !is Player) return
        val inventory = CustomInventory.get(event.inventory) ?: return

        if (!inventory.isOpen(event.whoClicked as Player)) {
            event.whoClicked.setMetadata("inventory", FixedMetadataValue(SantioUtils.plugin!!, inventory.id))
        }

        if (event.clickedInventory != inventory.getBukkitInventory()) return
        inventory.onClick(event)
    }

    @EventHandler
    private fun onQuit(event: PlayerQuitEvent) {
        SantioUtils.inventories.forEach { if (it.isOpen(event.player)) it.close(event.player) }
    }

}